Forms: fix invalid field ids

edanzer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #41527

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
For radio and multiple checkbox fields, when outputting options to the apage, we are using the option values for html id's. But option values can have spaces or odd characters, which produces invalid html ids. This PR uses the [santize_html_class()](https://developer.wordpress.org/reference/functions/sanitize_html_class/) method to strip white space and other non-valid characters. 

**Example of the issue:**
![invalid-html-ids](https://github.com/user-attachments/assets/d2b7b708-d15f-47f6-95e3-6b0b5ac1ec22)

**QUESTION:** Do we need to worry about backward compatibility issues when adjusting field ids. For now, my view is no. But the two likely concerns would be
(1) users applying custom CSS or JS to fields using the field ID
(2) changing the field IDS on existing forms -> as far as I can tell, those field ids are not used downstream for manipulating form submissions, but if we were or users are via custom code, that could have an impact. 

If we are concerned, we could consider wrapping the id output in conditionals - output the same as we have if the resulting value is valid, else output the new version (this new version would still be different, but if the old one was invalid, it's moot). 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None.

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add a form with a radio field and multiple checkbox field. Add radio and checkbox options with spaces or odd characters. 
* Go to the frontend, check the inspector, and confirm that the html id's on the fields are now valid (start with letter, no spaces or odd characters)
* Submit the form and check the form response in the dashboard (just to do a sanity check that form submissions work as expected)

edanzer committed:
Fix invalid radio/checkbox html ids

edanzer committed:
changelog

edanzer committed:
Update method docs

edanzer commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -745,7 +745,7 @@ private function render_consent_field( $id, $class ) {
 	 * Return the HTML for the multiple checkbox field.
 	 *
-	 * @param int    $id - the ID.
+	 * @param string $id - the ID (starts with 'g' - see constructor).
```
$id is always a string. See contructor around line 150 where it is set. This is relevant because a valid html ID need to start with a letter, not a number.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -674,11 +674,11 @@ public function render_radio_field( $id, $label, $value, $class, $required, $req
 			if ( is_string( $option ) && $option !== '' ) {
 				$radio_value = $this->get_option_value( $this->get_attribute( 'values' ), $option_index, $option );
-				$radio_id    = "$id-$radio_value";
+				$radio_id    = sanitize_html_class( $id ) . '-' . sanitize_html_class( $radio_value );
```
I don't think we need to wrap `$id` with sanitize_html_class( $id ). Since if this is the case we would need to do it in a few different places as well see line 668.

edanzer replied:
Well, if there's any chance that $id could contain non-valid characters for an html id, then the fact that we need to wrap it multiple times would be ok, and the right thing to do. 

But we create that id ourselves, and I believe it should be all valid characters, so you're right this is probably overkill for that reason.

enejb commented:
I didn't know about `sanitize_html_class` function. :) 

I wonder if something a bit different would do the trick mostly since we could have a case where we would produce the same ID and then we could take care of cases where a user could by accident produce the same ID. 
See
![Screenshot 2025-02-04 at 2 37 27 PM](https://github.com/user-attachments/assets/2a5d97c3-5c02-4684-a1ce-bd9e7f72009e)

Since the function produces the same ID we end up with the case where when you click on `OptionC` you select the `Option C` checkbox. 

Granted it is a bit of an edge case but `wp_unique_id()` could work well for solving the case instead.

edanzer commented:
@enejb @simison - As I read the code, a duplicate id should not be possible. The id is a [combination of two pieces](https://github.com/Automattic/jetpack/blob/6a78ad15d603a16b50d8cbcfdea03091e34603cd/projects/packages/forms/src/contact-form/class-contact-form-field.php#L677). The second part comes from the option name (which could be duplicated, as @enejb noted). 

But the first part is created in the constructor and has a check to specifically [avoid creating duplicates here](https://github.com/Automattic/jetpack/blob/6a78ad15d603a16b50d8cbcfdea03091e34603cd/projects/packages/forms/src/contact-form/class-contact-form-field.php#L159). 

Technically, we could just use that first part. I'm not sure why someone originally started appending the option name as well. I was just trying to keep the implementation here close to what we already had in place. But if we prefer, I'd also be fine if we want apply some kind of random unique ID as the second part if we prefer that.

edanzer committed:
Updated fix for duplicate form field ids

edanzer committed:
Remove unneeded sanitization

edanzer committed:
Add back esc_attr() calls

edanzer commented:
Implemented new solution here. 
- I considered wp_unique_id() but (1) it would give different ids on each page load and I think we'd prefer those be consistent and (2) using the option value is more semantic.
- I removed santize_html_field() on the id for the select/radio group - it is not needed.
- I'm still wrapping the option value in santize_html_field() to trim spaces and non-valid characters.
- When outputting options, we track the html ids and if there's a duplicate, append the option index. So given options like blue, blue, and red, the resulting html ids would like id-blue, id-blue-1, id-red.

edanzer commented:
I want to add a clarification note. At first glance, the extra code here to track ids and conditionally add an option index may seem needlessly compex. There are other ways to ensure unique id. We could do it with one line using id + option index, or using some unique id generator. 

The downsides to such approach are that you lose semantic value (readability) in the id, and you risk changing the ids for option in confusing ways. Take this example:

- A user has a radio field for 'Reason for Contact' that has two options, 'Tech Support' and 'Marketing'.
   - IDs like 'g159-tech-support' and 'g159-marketing' are arguably better than 'g159-1' and 'g159-2'.
- After two weeks, the user decides to add 'Feedback' as the first option.
   - Using option values, the user now has IDs of 'g159-feedback', 'g159-tech-support', and 'g159-marketing'. Clear and option IDs still match the same option as before.
   - Using just option index, the user now has 'g159-1', 'g159-2', and 'g159-3'. Unclear which is which, and g159-1 now refers to a different option than before (was 'Tech Support', now 'Feedback').
- After another two weeks, the user decides to switch the order to Marketing, Feedback, Tech Support.
   - With option values: 'g159-marketing', 'g159-feedback'', and 'g159-tech-support'. Still clear. Still the same.
   - With option index: 'g159-1', 'g159-2', 'g159-3'. We see the same option ids are there, but they all refer to different options than when we started.

It may be that not all form users need or care about the semantic readability or consistency in html id's, but some will. It would be especially useful if a user is trying to add custom CSS or JS to the form. And it doesn't cost us much in lines of code or complexity to keep readability and consistency.

I'd be fine adjusting, but that was the thinking behind the approach.

simison reviewed (approved):
I didn't test but code look good, and reasoning is good. 👍 

This did come up with Team 51 as pet-peeve problem with someone. :-)

enejb reviewed (approved):
Works as expected! Nicely done
